tests/e2e/tests/pilot/pod_churn_test.go

```diff
@@ -0,0 +1,281 @@
+//  Copyright 2018 Istio Authors
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package pilot
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/hashicorp/go-multierror"
+	"istio.io/istio/pkg/log"
+	"istio.io/istio/pkg/test/application/echo"
+	"istio.io/istio/pkg/test/application/echo/proto"
+	"istio.io/istio/pkg/test/kube"
+)
+
+const (
+	churnAppName         = "churn"
+	churnAppSelector     = "app=" + churnAppName
+	churnAppReplicas     = 10
+	churnAppMinReplicas  = 5
+	churnTrafficDuration = 1 * time.Minute
+
+	// churnTrafficThreads is the number of threads that will concurrently send traffic to the service being
+	// churned. This is set to Envoy's default circuit breaker parameter max_retry_attempts, which
+	// is the maximum number of parallel retries that Envoy will allow to the upstream cluster. If we
+	// go beyond that threshold, the circuit breaker itself will cause 503s.
+	churnTrafficThreads = 3
+)
+
+// TestPodChurn creates a replicated app and periodically brings down one pod. Traffic is sent during the pod churn
+// to ensure that no 503s are received by the application. This verifies that Envoy is properly retrying disconnected
+// endpoints.
+func TestPodChurn(t *testing.T) {
+	// Deploy the churn app
+	app := getApp(churnAppName,
+		churnAppName,
+		churnAppReplicas,
+		8080,
+		80,
+		9090,
+		90,
+		7070,
+		70,
+		"v1",
+		true,
+		false,
+		true,
+		true)
+	if err := tc.Kube.AppManager.DeployApp(&app); err != nil {
+		t.Fatal(err)
+	}
+	defer func() {
+		_ = tc.Kube.AppManager.UndeployApp(&app)
+	}()
+	fetchFn := tc.Kube.KubeAccessor.NewPodFetch(tc.Kube.Namespace, churnAppSelector)
+	if err := tc.Kube.KubeAccessor.WaitUntilPodsAreReady(fetchFn); err != nil {
+		t.Fatal(err)
+	}
+
+	// Start a pod churner on the app. The pod churner will periodically bring down one of the pods for the app.
+	podChurner := newPodChurner(tc.Kube.Namespace, churnAppSelector, churnAppMinReplicas, tc.Kube.KubeAccessor)
+	podChurner.Start()
+	defer func() {
+		_ = podChurner.Close()
+	}()
+
+	// Create a port forwarder so that we can send commands app "a" to talk to the churn app.
+	forwarder, err := newPortForwarder(primaryCluster, "a", 7070)
```

Comment:
why do you need this? In other tests, we simply send traffic from one of the pods in the cluster (like pod t, using the default client app that is capable of sending concurrent requests.). Then similar to the other routing tests, you can simply count the response codes to make sure that there are no 5xx or failures.

Reply:
I'm doing the same thing, but in a different way. The old way used exec which is slow.  Since I'm re-sending requests every second, I'm using a persistent gRPC connection to the source pod to control the traffic, rather than doing exec.

Reply:
ah nice.. Thanks.